### ISSUE TITLE
Core refactor: WorkerInfrastructure Task optimization

### ISSUE THREAD

**Issue by thehapax** [Codaone/DEXBot#589]
multithreading or processing

**Comment by thehapax** [comment]
worker infrastructure has redundancy in using config; reduce. 
propose 

- 1 worker = 1 strategy
- multiple threads when using worker infrastructure
- can also be extended to multi process

**Comment by bitfag** [comment]
>1 worker = 1 strategy

Same as current.

> multiple threads when using worker infrastructure

So you decided to not use asyncio? Can you post wider description why such decision has been made?

**Comment by thehapax** [comment]
Not final decision. right now just experimenting.

**Comment by thehapax** [comment]
It appears that the current implementation is 1 WorkerInfrastructure implements multiple StrategyBase, unless i'm reading this incorrectly?

**Comment by bitfag** [comment]
WorkerInfrastructure is a supervisor which instantiates workers, 1 worker = 1 Strategy instance.

**Comment by thehapax** [comment]
Ok so tell me if i am getting this wrong, because i have not implemented asyncio with multithreading. I did a little reading and some of it is still confusing, so here is my best take: 

Here is what i see so far: 
I see a lot of redundant loops calling the config file inside of WorkerInfrastructure. Why not make it simpler, and just handle the config file parsing outside the class?

Afaik, WorkerInfrastructure is currently threaded with locks because of the config file. I think we can get rid of the threading and locks for config file by handling the creation of all the WorkerInfrastructure objects (each with only 1 worker in them) and all of what's done in: init_workers(self, config), outside of the class and put into a list_of_workers, then: 

once all the workers are created and put into list_of_workers, we can use a ThreadPool Executor 
```
futures = []
        with ThreadPoolExecutor(MAX_WORKERS) as executor:
            for obj in list_of_workers:
                futures.append(executor.submit(obj.run))
```
I know the above is just regular threads, not async, but I'm assuming we'd want to do something like : https://docs.python.org/3/library/asyncio-eventloop.html#executing-code-in-thread-or-process-pools 

and modify the run to handle async for update_notify, listen, onMarketUpdate, onAccount, etc. as it appears that these along with run are the methods that are blocking:

```
 def run(self):
        self.update_notify()
        self.notify.listen()
```

This way we can just async all the necessary methods in the WorkerInfrastructure class and the new async bitshares.notify

Or perhaps this Queue model would be better? https://realpython.com/async-io-python/#using-a-queue

What do you think?

**Comment by thehapax** [comment]
One other thing - it appears that switching between process pools and thread pools is not much of a big deal. see link above from asyncio-eventloop.html. 

I am aware that most users who are running in the cloud would be on 1 cpu. However, for us power desktop users with multicore, would be kind of nice to have the multiprocessing option. I have 4 cores on my laptop and a 12 core desktop....

**Comment by joelvai** [comment]
It is true that WorkerInfrastructure is the central place and that 1 worker has only 1 strategy like @bitfag said, but the thing is that WorkerInfrastructure is running on single thread.

**Comment by bitfag** [comment]
> I see a lot of redundant loops calling the config file inside of WorkerInfrastructure. Why not make it simpler, and just handle the config file parsing outside the class?

I didn't thought about this, looks like a good idea.

Further:

* Multiprocessing make sense when you have a CPU-greedy code. This may be signing of thousands of transactions or some heavy computation
* Mixing asyncio and ThreadPoolExecutor makes sense if you have a synchronous legacy code you want to run in parallel not blocking the event loop.
* Threads can have shared variables etc, and when you switching to multiprocessing data exchange between these processes become more tricky

The whole discussion doubles a previous one we have, where I described my thoughts on architecture here https://github.com/Codaone/DEXBot/issues/538#issuecomment-480501436

**Comment by thehapax** [comment]
I know that we have #538, but I want to make sure we agree here on the worker.py refactor which split into this issue. The #538 has too many points in it and want to separate it out. 

Based on discussion above, until the async part of bitshares.notify is done, I just want to make clear to you what I am doing with ThreadPoolExecutor and put it on a branch until the async in pybitshares is ready. After we can make changes accordingly.

**Comment by thehapax** [comment]
I made a test branch here for ThreadPool, with a simplified core_worker and core_cli so that it doesn't interfere with existing code. i hope this is the right direction

https://github.com/thehapax/DEXBot/tree/core-refactor-test

**Comment by bitfag** [comment]
I did quick review of this branch. I see that all workers are using single `BitShares` instance. The problem it's not threadsafe, and I got into situation when both workers tried to broadcast in parallel, resulting in `grapheneapi.exceptions.RPCError: missing required active authority: Missing Active Authority ` error.

Further, I think there may be conflicts with subscriptions as each `CoreWorkerInfrastructure` manages own Notify. I suspect they are overriding each other subscriptions (I didn't checked it, just thinking).

**Comment by thehapax** [comment]
One possible way to resolve the bitshares instance is make it threadsafe, as it was before with locks. I removed the locks as it seemed it was only used for the config file, so this is an easy fix. 

If subscriptions also do ok with locks then I will pass this branch on to bitProfessor for additional testing. 

In the meanwhile will wait for your async pybitshares methods to complete before making core worker async, assuming that async will change notify calls.

**Comment by bitfag** [comment]
These locks were for config only, python-bitshares itself should be refactored to be threadsafe.

**Comment by thehapax** [comment]
Ok, so will you refactor python-bitshares threadsafe?

**Comment by bitfag** [comment]
I currently working on asyncio support at library level, I would like to test some concepts here and bring this at MVP state, then I can investigate thread safety there.

**Comment by bitfag** [comment]
Still having issues with multiple workers. You're creating own `self.event_lock` for each thread, so locking doesn't actually work. You need to use the same RLock object across all threads.

**Comment by bitfag** [comment]
The error means both threads modified bitshares instance txbuffer.

```
2019-05-27 21:23:46,074 (bitshares_feed:319) - dtest2 - DEBUG - Center price in get_market_center_price: 0.99999752 
2019-05-27 21:23:46,449 (staggered_orders:575) - dtest2 - DEBUG - Need to allocate base: 416.2606872 GOVNO
2019-05-27 21:23:46,450 (staggered_orders:661) - dtest2 - DEBUG - Placing closer buy order; actual spread: 6.1516%, target + increment: 2.5000%
2019-05-27 21:23:46,642 (staggered_orders:1378) - dtest2 - INFO - Placing closer buy order
2019-05-27 21:23:46,643 (bitshares_engine:550) - dtest2 - INFO - Placing a buy order with 103.2877660 GOVNO @ 0.97544262
2019-05-27 21:23:46,739 (bitshares_feed:319) - dtest1 - DEBUG - Center price in get_market_center_price: 1.00001254 
2019-05-27 21:23:47,166 (staggered_orders:575) - dtest1 - DEBUG - Need to allocate base: 399.7446125 GOVNO
2019-05-27 21:23:47,167 (staggered_orders:661) - dtest1 - DEBUG - Placing closer buy order; actual spread: 6.1548%, target + increment: 2.5000%
2019-05-27 21:23:47,261 (staggered_orders:1378) - dtest1 - INFO - Placing closer buy order
2019-05-27 21:23:47,261 (bitshares_engine:550) - dtest1 - INFO - Placing a buy order with 99.1537427 GOVNO @ 0.97544262
2019-05-27 21:23:48,203 (bitshares_engine:565) - dtest2 - DEBUG - Placed buy order {'expiration': '2019-05-27T16:24:48', 'ref_block_num': 47977, 'ref_blo
ck_prefix': 1304693786, 'operations': [[1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23299', 'amount_to_sell': {'amount': 1032877661, 
'asset_id': '1.3.1070'}, 'min_to_receive': {'amount': 1058881, 'asset_id': '1.3.1018'}, 'expiration': '2024-05-25T16:23:46', 'fill_or_kill': False, 'exte
nsions': []}], [1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23242', 'amount_to_sell': {'amount': 991537427, 'asset_id': '1.3.1070'}, 
'min_to_receive': {'amount': 10165, 'asset_id': '1.3.1271'}, 'expiration': '2024-05-25T16:23:47', 'fill_or_kill': False, 'extensions': []}]], 'extensions
': [], 'signatures': []}
2019-05-27 21:23:48,213 (staggered_orders:563) - dtest2 - DEBUG - Storing balance data at center price 0.99999752
2019-05-27 21:23:48,225 (staggered_orders:575) - dtest2 - DEBUG - Need to allocate quote: 61.2008 GOLODRANETS
2019-05-27 21:23:48,238 (staggered_orders:661) - dtest2 - DEBUG - Placing closer sell order; actual spread: 6.1516%, target + increment: 2.5000%
2019-05-27 21:23:48,342 (staggered_orders:1387) - dtest2 - INFO - Placing closer sell order
2019-05-27 21:23:48,352 (bitshares_engine:606) - dtest2 - INFO - Placing a sell order with 15.1857 GOLODRANETS @ 1.02517054
2019-05-27 21:23:48,580 (bitshares_engine:621) - dtest2 - DEBUG - Placed sell order {'expiration': '2019-05-27T16:24:48', 'ref_block_num': 47977, 'ref_bl
ock_prefix': 1304693786, 'operations': [[1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23299', 'amount_to_sell': {'amount': 1032877661,
 'asset_id': '1.3.1070'}, 'min_to_receive': {'amount': 1058881, 'asset_id': '1.3.1018'}, 'expiration': '2024-05-25T16:23:46', 'fill_or_kill': False, 'ext
ensions': []}], [1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23242', 'amount_to_sell': {'amount': 991537427, 'asset_id': '1.3.1070'},
 'min_to_receive': {'amount': 10165, 'asset_id': '1.3.1271'}, 'expiration': '2024-05-25T16:23:47', 'fill_or_kill': False, 'extensions': []}], [1, {'fee':
 {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23299', 'amount_to_sell': {'amount': 151858, 'asset_id': '1.3.1018'}, 'min_to_receive': {'amount': 
155679846, 'asset_id': '1.3.1070'}, 'expiration': '2024-05-25T16:23:48', 'fill_or_kill': False, 'extensions': []}]], 'extensions': [], 'signatures': []}
2019-05-27 21:23:48,582 (staggered_orders:563) - dtest2 - DEBUG - Storing balance data at center price 0.99999752
2019-05-27 21:23:48,676 (bitshares_engine:565) - dtest1 - DEBUG - Placed buy order {'expiration': '2019-05-27T16:24:48', 'ref_block_num': 47977, 'ref_block_prefix': 1304693786, 'operations': [[1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23299', 'amount_to_sell': {'amount': 1032877661, 'asset_id': '1.3.1070'}, 'min_to_receive': {'amount': 1058881, 'asset_id': '1.3.1018'}, 'expiration': '2024-05-25T16:23:46', 'fill_or_kill': False, 'extensions': []}], [1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23242', 'amount_to_sell': {'amount': 991537427, 'asset_id': '1.3.1070'}, 'min_to_receive': {'amount': 10165, 'asset_id': '1.3.1271'}, 'expiration': '2024-05-25T16:23:47', 'fill_or_kill': False, 'extensions': []}], [1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23299', 'amount_to_sell': {'amount': 151858, 'asset_id': '1.3.1018'}, 'min_to_receive': {'amount': 155679846, 'asset_id': '1.3.1070'}, 'expiration': '2024-05-25T16:23:48', 'fill_or_kill': False, 'extensions': []}]], 'extensions': [], 'signatures': []}
2019-05-27 21:23:48,678 (staggered_orders:563) - dtest1 - DEBUG - Storing balance data at center price 1.00001254
2019-05-27 21:23:48,678 (staggered_orders:575) - dtest1 - DEBUG - Need to allocate quote: 757.88 PREC2
2019-05-27 21:23:48,679 (staggered_orders:661) - dtest1 - DEBUG - Placing closer sell order; actual spread: 6.1548%, target + increment: 2.5000%
2019-05-27 21:23:48,872 (staggered_orders:1387) - dtest1 - INFO - Placing closer sell order
2019-05-27 21:23:48,873 (bitshares_engine:606) - dtest1 - INFO - Placing a sell order with 188.00 PREC2 @ 1.02520133
2019-05-27 21:23:49,259 (bitshares_engine:621) - dtest1 - DEBUG - Placed sell order {'expiration': '2019-05-27T16:24:48', 'ref_block_num': 47977, 'ref_block_prefix': 1304693786, 'operations': [[1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23299', 'amount_to_sell': {'amount': 1032877661, 'asset_id': '1.3.1070'}, 'min_to_receive': {'amount': 1058881, 'asset_id': '1.3.1018'}, 'expiration': '2024-05-25T16:23:46', 'fill_or_kill': False, 'extensions': []}], [1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23242', 'amount_to_sell': {'amount': 991537427, 'asset_id': '1.3.1070'}, 'min_to_receive': {'amount': 10165, 'asset_id': '1.3.1271'}, 'expiration': '2024-05-25T16:23:47', 'fill_or_kill': False, 'extensions': []}], [1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23299', 'amount_to_sell': {'amount': 151858, 'asset_id': '1.3.1018'}, 'min_to_receive': {'amount': 155679846, 'asset_id': '1.3.1070'}, 'expiration': '2024-05-25T16:23:48', 'fill_or_kill': False, 'extensions': []}], [1, {'fee': {'amount': 100, 'asset_id': '1.3.0'}, 'seller': '1.2.23242', 'amount_to_sell': {'amount': 18801, 'asset_id': '1.3.1271'}, 'min_to_receive': {'amount': 1927433347, 'asset_id': '1.3.1070'}, 'expiration': '2024-05-25T16:23:49', 'fill_or_kill': False, 'extensions': []}]], 'extensions': [], 'signatures': []}
2019-05-27 21:23:49,260 (staggered_orders:563) - dtest1 - DEBUG - Storing balance data at center price 1.00001254
2019-05-27 21:23:49,565 (staggered_orders:241) - dtest2 - ERROR - Got exception during broadcasting trx:
Traceback (most recent call last):
  File "/home/vvk/devel/bitshares/python-graphenelib/grapheneapi/api.py", line 168, in func
    r = func(*args, **kwargs)
  File "/home/vvk/devel/bitshares/python-graphenelib/grapheneapi/rpc.py", line 138, in method
    message = self.parse_response(r)
  File "/home/vvk/devel/bitshares/python-graphenelib/grapheneapi/rpc.py", line 106, in parse_response
    raise RPCError(ret["error"]["message"])
grapheneapi.exceptions.RPCError: missing required active authority: Missing Active Authority 1.2.23242

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/vvk/devel/DEXBot/dexbot/strategies/staggered_orders.py", line 236, in maintain_strategy
    self.execute()
  File "/home/vvk/devel/DEXBot/dexbot/orderengines/bitshares_engine.py", line 476, in execute
    r = self.bitshares.txbuffer.broadcast()
  File "/home/vvk/devel/bitshares/python-graphenelib/graphenecommon/transactionbuilder.py", line 504, in broadcast
    raise e
  File "/home/vvk/devel/bitshares/python-graphenelib/graphenecommon/transactionbuilder.py", line 498, in broadcast
    ret, api="network_broadcast"
  File "/home/vvk/devel/bitshares/python-graphenelib/grapheneapi/api.py", line 176, in func
    self.post_process_exception(e)
  File "/home/vvk/devel/bitshares/python-bitshares/bitsharesapi/bitsharesnoderpc.py", line 24, in post_process_exception
    raise exceptions.UnhandledRPCError(msg)
bitsharesapi.exceptions.UnhandledRPCError: missing required active authority: Missing Active Authority 1.2.23242
2019-05-27 21:23:49,762 (core_worker:125) - dtest1 - ERROR - in ontick() dtest1 
Traceback (most recent call last):
  File "/home/vvk/devel/DEXBot/dexbot/core_worker.py", line 121, in on_block
    self.workers[self.worker_name].ontick(data)
  File "/home/vvk/.local/share/virtualenvs/DEXBot-XTl2tJdV/lib/python3.6/site-packages/Events-0.3-py3.6.egg/events/events.py", line 95, in __call__
    f(*a, **kw)
  File "/home/vvk/devel/DEXBot/dexbot/strategies/staggered_orders.py", line 1855, in tick
    self.maintain_strategy()
  File "/home/vvk/devel/DEXBot/dexbot/strategies/staggered_orders.py", line 236, in maintain_strategy
    self.execute()
  File "/home/vvk/devel/DEXBot/dexbot/orderengines/bitshares_engine.py", line 476, in execute
    r = self.bitshares.txbuffer.broadcast()
  File "/home/vvk/devel/bitshares/python-graphenelib/graphenecommon/transactionbuilder.py", line 478, in broadcast
    self.sign()
  File "/home/vvk/devel/bitshares/python-graphenelib/graphenecommon/transactionbuilder.py", line 456, in sign
    raise MissingKeyError
graphenecommon.exceptions.MissingKeyError
2019-05-27 21:23:49,764 (core_worker:118) - dtest1 - ERROR - Worker "dtest1" is disabled
```

If you'll begin to use shared RLock, only one worker will run at a time, diminishing any threading advantages. This will be basically the same what we have now, but in threaded manner. Proper multithread handling should be performed inside python-bitshares.

**Comment by thehapax** [comment]
Is there a possibility of using a separate bitshares instance for each worker, and have each on its on process. I do not see any need for each worker to need shared information with each other than the bitshares instance, so they can be in separate processes. 

Also Inside of RO, there are todo items that have possible race condition
```
                 # Detect partially filled orders;
                    # on fresh order 'for_sale' is always equal to ['base']['amount']
                    if current_order['for_sale']['amount'] != current_order['base']['amount']:
                        diff_abs = current_order['base']['amount'] - current_order['for_sale']['amount']
                        diff_rel = diff_abs / current_order['base']['amount']
                        if diff_rel >= self.partial_fill_threshold:
                            need_update = True
                            self.log.info('Partially filled order detected, filled {:.2%}'.format(diff_rel))
                            # FIXME: Need to write trade operation; possible race condition may occur: while
                            #        we're updating order it may be filled further so trade log entry will not
                            #        be correct
```

and in SO, maintain_strategy() race condition. 
```
# Send pending operations
        trx_executed = False
        if not self.bitshares.txbuffer.is_empty():
            trx_executed = True
            try:
                self.execute()
            except bitsharesapi.exceptions.RPCError:
                """ Handle exception without stopping the worker. The goal is to handle race condition when partially
                    filled order was further filled before we actually replaced them.
                """
                self.log.exception('Got exception during broadcasting trx:')
                return
        self.bitshares.bundle = False
```

**Comment by thehapax** [comment]
RLock vs Lock is not making much of a performance difference for processes, i just tested

**Comment by thehapax** [comment]
oh wait, can't have multiple bitshares instances because of wallet locking. hm.

**Comment by thehapax** [comment]
structurally, even if we have thread safe python-bitshares, would we still need to do signaling between threads at the worker level for events, like this?
https://pymotw.com/3/threading/index.html#signaling-between-threads

Please let me know your thoughts. i am putting this on hold until next week while i work on other tasks.

**Comment by MarkoPaasila** [comment]
I can't think of any reason workers would need to communicate. One would be if we wanted to make sure strategies aren't trading against each other, but we shouldn't have that concern as we don't take orders. That would also become hugely complicated...

**Comment by bitfag** [comment]
>Also Inside of RO, there are todo items that have possible race condition

These race conditions are not the result of issues in code, they are just possible because something may happen on the bitshares side between querying orders and putting new orders for example.

>would we still need to do signaling between threads at the worker level

I think no signaling required between workers as they are independent. Though it's possible that in the future some small signaling may be needed for case when multiple workers running on single account with assets intersections.

**Comment by thehapax** [comment]
@MarkoPaasila The one issue I have with RO right now is that in the init() is it calling update_orders() or check_orders().  for one, it makes for a very slow initialization start, but these order updating logic should be handled outside of init(), as in SO, which does all the logic in maintain_strategy(). SO Initialization is also much faster. I'd like to fix this.
 
as in the case mentioned before with separate Bitshares instances, and running more than one worker, for RO, the init() also creates thread collision issues, even if a separate thread with locks.  the time for the check/update orders method to return is also really slow.

**Comment by thehapax** [comment]
Fwiw, I am not sure why we have issues with losing bitshares instance. Would be great to get some clarification on why we have to keep adding this parameter to more and more methods in dexbot. 

Example: 
Asset(fee_asset, bitshares_instance=self.bitshares)
order = Order(order_id, blockchain_instance=self.bitshares)

https://github.com/Codaone/DEXBot/pull/558

**Comment by bitfag** [comment]
>  bitshares_instance=self.bitshares

This is not strictly needed as we setting `shared_bitshares_instance()`. It was a library problem in some early 3.x version, so shared instance didn't worked.
